File: api/src/toornament.py
import requests
import json

class ApiUser:
    def __init__(self, api):
        self.api = api
        self.access_token = None

    def get_tokens(self, code):
        r = requests.post("https://api.toornament.com/oauth/v2/token", data = {
            "grant_type": "authorization_code",
            "client_id": self.api.client_id,
            "client_secret": self.api.client_secret,
            "redirect_uri": self.api.redirect_uri,
            "code": code
        })

        if r.status_code != 200:
            return None

        return r.json()

    def get_tournaments(self):
        r = requests.get("https://api.toornament.com/organizer/v2/tournaments?disciplines=player_unknowns_battlegrounds", headers = {
            "X-Api-Key": self.api.api_key,
            "Authorization": self.access_token,
            "Range": "tournaments=0-19"
        })

        print(r)
        print(r.text)

        return r.json()
    
    def get_matches(self, tournament_id):
        r = requests.get("https://api.toornament.com/organizer/v2/tournaments/{}/matches".format(tournament_id), headers = {
            "X-Api-Key": self.api.api_key,
            "Authorization": self.access_token,
            "Range": "matches=0-49"
        })

        print(r)
        print(r.text)

        return r.json()

    def get_games(self, tournament_id, match_id):
        r = requests.get("https://api.toornament.com/organizer/v2/tournaments/{}/matches/{}/games".format(tournament_id, match_id), headers = {
            "X-Api-Key": self.api.api_key,
            "Authorization": self.access_token,
            "Range": "games=0-49"
        })

        print(r)
        print(r.text)

        return r.json()

    def get_match(self, tournament_id, match_id):
        r = requests.get("https://api.toornament.com/organizer/v2/tournaments/{}/matches/{}".format(tournament_id, match_id), headers = {
            "X-Api-Key": self.api.api_key,
            "Authorization": self.access_token
        })

        print(r)

        return r.json()

    def patch_toornament_game(self, tournament_id, match_id, game_number, data):
        r = requests.patch("https://api.toornament.com/organizer/v2/tournaments/{}/matches/{}/games/{}".format(tournament_id, match_id, game_number), headers = {
            "X-Api-Key": self.api.api_key,
            "Authorization": self.access_token
        }, data=json.dumps(data))
        
        print(r)

class Api:
    def __init__(self, api_key, client_id, client_secret, redirect_uri):
        self.api_key = api_key
        self.client_id = client_id
        self.client_secret = client_secret
        self.redirect_uri = redirect_uri
        self.token = None

File: api/src/toornament.rs
use reqwest::Client;

#[derive(Debug)]
pub struct ToornamentApi {
    pub api_uri: String,
    pub api_key: String,
    pub client_id: String,
    pub client_secret: String,
    pub redirect_uri: String
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum TokenResponse {
    TokenResult {
        scope: String,
        token_type: String,
        expires_in: u64,
        access_token: String,
        refresh_token: String
    },
    TokenError {
        error: String,
        message: String,
        hint: Option<String>
    }
}

pub trait ToornamentAuthorizer {
    fn authorize_with_code(&mut self, code: String) -> Result<TokenResponse, String>;
}

impl ToornamentAuthorizer for ToornamentApi {
    fn authorize_with_code(&mut self, code: String) -> Result<TokenResponse, String> {
        let client = Client::new();

        let params = [
            ("grant_type", "authorization_code"),
            ("client_id", &format!("{}", self.client_id)),
            ("client_secret", &format!("{}", self.client_secret)),
            ("redirect_uri", &format!("{}", self.redirect_uri)),
            ("code", code.as_str())
        ];

        let res = client.post(&format!("{}{}", self.api_uri, "/oauth/v2/token"))
            .form(&params)
            .send();

        if res.is_err() {
            println!("{:?}", res);

            return Err("ko".to_string());
        }

        match res.unwrap().json::<TokenResponse>() {
            Ok(tokens)  => Ok(tokens),
            Err(e)      => Err(e.to_string())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use std::error::Error;
    use std::fs::File;
    use std::io::Read;

    use serde_json::*;

    #[test]
    fn internal() {
        let mut file = File::open("fixtures/token_ok.json").unwrap();
        let mut contents = String::new();
        file.read_to_string(&mut contents).unwrap();

        let data: TokenResponse = serde_json::from_str(&contents).unwrap();
        println!("{:#?}", data);
    }
}
